### DISCUSSION TITLE
Tagging metrics with concourse build information

### DISCUSSION THREAD

**Discussion by mdbenjam** [concourse/concourse#8288] [Help & Support] [answered]
Hi all,

We run our test suite in a concourse task, and in that task we emit metrics to [DataDog's CI tool](https://docs.datadoghq.com/continuous_integration/). We'd like to include the URL of the job as a tag on the metrics we send to DataDog so that we can group test runs by build, and have a link between the builds and our monitoring infrastructure. I'm wondering what the best way to do this is since the [build env vars are not exposed to tasks](https://github.com/concourse/concourse/issues/494). Is it possible to use a resource for this since the metrics are sent to the DataDog agent while the tests are running.

I'd appreciate any guidance!

**Comment by suhlig** [comment, marked as answer]
My way of thinking about Concourse is that everything that has an effect on external systems should be a resource instead of a task. A neat side-effect is that the web UI visualizes the output with line coming out of a job.

If you rewrite the Datadog task as `put` step of a resource, [the information you want is available](https://concourse-ci.org/implementing-resource-types.html#resource-metadata) to your resource. I find the [phraseapp resource](https://github.com/tenjaa/concourse-phraseapp-resource) to be a nice boilerplate.

**Reply by mdbenjam** [reply]
Thanks for the answer, this helps me better understand the difference between tasks and resources.

In our case this is a little challenging because our task, which runs our tests, uses DataDog's CI integration to automatically send metrics from within the task as tests pass or fail. This might be hard to change, but it sounds like the Concourse-y way to do it would be to package all these updates into an artifact of some sort, and pass it to a resource that uploads all of the DataDog information with the tags we want. This architecture would make our task side-effect free, and would rely on our resource for interaction with the outside world.

Until we have time to split out these side effects from our tasks, I've used the [Smuggler Resource](https://github.com/redfactorlabs/concourse-smuggler-resource) to make the build information available to our task. But I recognize this is breaking out of the intended design of tasks and resources in concourse 😢 .
